_posts/2020-10-23-deconvolution.md

```diff
@@ -0,0 +1,387 @@
+  <line x1="130" y1="0" x2="162" y2="32" style="stroke-width:2" />
+
+  <!-- Colored Rectangle -->
+  <polygon points="10.000000,0.000000 130.000000,0.000000 162.743566,32.743566 42.743566,32.743566" style="fill:#ECB172A0;stroke-width:0"/>
+
+  <!-- Horizontal lines -->
+  <line x1="42" y1="32" x2="162" y2="32" style="stroke-width:2" />
+  <line x1="42" y1="152" x2="162" y2="152" style="stroke-width:2" />
+
+  <!-- Vertical lines -->
+  <line x1="42" y1="32" x2="42" y2="152" style="stroke-width:2" />
+  <line x1="162" y1="32" x2="162" y2="152" style="stroke-width:2" />
+
+  <!-- Colored Rectangle -->
+  <polygon points="42.743566,32.743566 162.743566,32.743566 162.743566,152.743566 42.743566,152.743566" style="fill:#ECB172A0;stroke-width:0"/>
+
+  <!-- Text -->
+  <text x="102.743566" y="172.743566" font-size="1.0rem" font-weight="100" text-anchor="middle" >2048</text>
+  <text x="182.743566" y="92.743566" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,182.743566,92.743566)">2048</text>
+  <text x="16.371783" y="156.371783" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,16.371783,156.371783)">950</text>
+</svg>
+</td>
+</tr>
+</table>
+
+
+What we now have is a Dask array composed of 16 CuPy blocks of data.  Notice how Dask provides nice typing information in the SVG output.  When we moved from NumPy to CuPy, the block diagram above displays `Type: cupy.ndarray` -- these are especially helpful and remind us we are operating on data which lives on the GPU.
+
+The last piece we need before running the deconvolution is the PSF which should also be loaded onto the GPU:
+
+```python
+avg_psf = skimage.io.imread("/public/NVMICROSCOPY/AVG_PSF.tif")
+c_avg_psf = cp.asarray(avg_psf)
+```
+
+Lastly, we map the `deconvolve` function onto each block in our Dask array:
+
+```
+out = da.map_blocks(deconvolve,
+                    c_imgs,
+                    c_psf,
+                    iterations=20,
+                    dtype=cp.float32)
+```
+
+With Dask and multiple GPUs, we measured deconvolution of an 8GB image in ~10 seconds! But this is just the first step towards accelerated image science.
+
+<a href="/images/deconvolve.png">
+    <img src="/images/deconvolve.png" width="100%"></a>
+
+IMAGE DESCRIPTION
+
+Napari + Remote GPUs
+--------------------
+
+Deconvolution is just one operation and one tool, an image scientist or microscopist will need.  They will need other tools to help as the push to understand
+more of the underlying biology.  As a first step though, they will need tools to visualize the data. [Napari](https://napari.org/) is a multi-dimensional image viewer popular in the PyData Bio ecosystem.  As an experiment, we extended the demo above to not only work with Napari, but to also work on remote GPUs provided
+by [coiled.io](https://coiled.io/) [Example Notebook](https://gist.github.com/quasiben/29901204b20421b18f029e1fd07ed0cf)
```

Comment:
It might be good to add a line saying that you could use other remote services like X or Y, but we picked Coiled because $reason

Reply:
Using other services seems fine. My guess is any cloud or local resource would work. There's some setup effort incurred that varies depending on the scale, location, and what might already have been done to deploy there. Should we say something like this?

I don't know that we had a specific reason. It was easy. We were able to bug Matt to help us with things 😉. Is that worth mentioning? 🙂

Reply:
I tried rewording this and filling in other options. Please let us know what you think 😄